kotlin/authenticate-with-oauth/src/main/java/com/esri/arcgisruntime/sample/authenticatewithoauth/MainActivity.kt

```diff
@@ -0,0 +1,348 @@
+          }
+
+          // OAUTH_CREDENTIAL_CHALLENGE is issued when the portal is detected as supporting OAuth
+          if (authChallenge.type == AuthenticationChallenge.Type.OAUTH_CREDENTIAL_CHALLENGE) {
+            // attempt to use stored access token for auth
+            return AuthenticationChallengeResponse(
+              AuthenticationChallengeResponse.Action.CONTINUE_WITH_CREDENTIAL,
+              UserCredential.createFromToken(accessToken, authChallenge.remoteResource?.uri)
+            )
+          }
+        }
+
+        // if SharedPreferences has an auth code, we've likely just been through the OAuth flow and now have an auth code
+        // we can use to request a new access token
+        sharedPreferences.authCode?.let {
+          // use the authorization code to get a new access token by executing an OAuthTokenCredentialRequest
+          val request = OAuthTokenCredentialRequest(
+            oAuthConfig.portalUrl,
+            null,
+            oAuthConfig.clientId,
+            oAuthConfig.redirectUri,
+            sharedPreferences.authCode
+          )
+
+          val credential = request.executeAsync().get()
+          with(sharedPreferences) {
+            putAccessToken(credential.accessToken)
+            putAccessTokenExpiry(System.currentTimeMillis() + (ACCESS_TOKEN_EXPIRY_MINS * 60))
+            clearAuthCode()
+          }
+          // continue with credentials generated using auth code
+          return AuthenticationChallengeResponse(
+            AuthenticationChallengeResponse.Action.CONTINUE_WITH_CREDENTIAL,
+            credential
+          )
+        }
+
+        // we only want to begin OAuth here if the user has yet to authorize successfully
+        if (authChallenge.failureCount < 2) {
+          beginOAuth()
+        }
+
+        return AuthenticationChallengeResponse(AuthenticationChallengeResponse.Action.CANCEL, null)
+      } catch (e: Exception) {
+        // auth code has likely expired, clear existing auth code and begin OAuth flow
+        getString(R.string.error_auth_exception, e.message).let {
+          Log.d(TAG, it)
+          runOnUiThread {
+            logToUser(it)
+          }
+        }
+        sharedPreferences.clearAuthCode()
+        beginOAuth()
+      }
+    }
+    return AuthenticationChallengeResponse(AuthenticationChallengeResponse.Action.CANCEL, null)
+  }
+
+  private fun beginOAuth() {
+    // get the authorization code by sending user to the authorization screen
+    val authorizationUrl = OAuthTokenCredentialRequest.getAuthorizationUrl(
+      oAuthConfig.portalUrl, oAuthConfig.clientId, oAuthConfig.redirectUri, ACCESS_TOKEN_EXPIRY_MINS
+    )
+
+    // create an Intent to attempt to handle the authorization URL
+    with(Intent(Intent.ACTION_VIEW, Uri.parse(authorizationUrl))) {
+      this.resolveActivity(packageManager)?.let {
+        // this identifier ensures that the browser will attempt to reuse the same window each time the application
```

Comment:
I am not clear on what is the intent of this sample. The java version that worked on for ArcGIS indoors team was to illustrate how to use custom challenge handler to use OAuth in a WebView. 

If the use wants to launch the login window in a separate browser, they can just use the `DefaultChallengeHandler`.  Or if they want to preemptively challenge the user they can use `OAuthLoginManager.launchOAuthBrowserPage​()` 
https://developers.arcgis.com/android/latest/api-reference/reference/com/esri/arcgisruntime/security/OAuthLoginManager.html

Reply:
Double check with Xeuming but I believe the intent was to show usage a of Custom Challenge Handler but as Xeuming pointed out before, the use of a `WebView` is a security issue. So I showed how to use with a web browser. Also, I believe they wanted it in Kotlin.

Reply:
My Point is that this is a Sample, which our users use as a starting point. Most will copy/paste this code. Here we are not using our API that we provide. We have an issue to fix our API with this call to launch the same browser window. We should be making calls to our API here and not showing the user how to launch the browser window in Android. 

Also This logic here to launch the browser window does not make sense, as I don't think there any modern phones out there that are shipped without a web browser. So why don't we just restrict this sample to just show the use of Webview and have a separate Sample that shows the use of DefaultChallengeHandler that dishes out the OAuth prompt in a web browser.

Reply:
We have separate workflows for dishing out the challenge in the Browser via the DefaultChallangeHandler and using the WebView. So that we can handle the OAuth challenge from the point we get the Authorization code back from the browser. In the case of DefaultChallengeHandler, the user will have to specify the `DefaultOAuthIntentReceiver` activity so that the ArcGIS api recieves the authorization code and can request OAuth credential. In the case of WebView the control goes back to the shouldOverrideUrlLoading() where they can request OAuthCredential

Reply:
We aren't using `DefaultChallengeHandler`, this sample is for a custom Challenge Handler.